crates/autopilot/src/domain/competition/mod.rs

```diff
@@ -91,6 +91,35 @@ impl Score {
     }
 }
 
+#[derive(Clone)]
+pub struct RawParticipant {
```

Comment:
Not sure we gain much from this extra type. Instead we could initialize `Participant` with `winner: false` and have a second step that sets the value to `true` for the winners.

Reply:
In general, using the same struct to keep two or more different states of the object is an invitation for bugs to join the party.

If code duplication is issue here, I would rather have 

```
#[derive(Clone)]
pub struct Participant {
    inner: RawParticipant,
    winner: bool,
}

impl Participant {
    pub fn driver(&self) -> &Arc<infra::Driver> {
        &self.inner.driver
    }

    pub fn solution(&self) -> &Solution {
        &self.inner.solution
    }

    pub fn is_winner(&self) -> bool {
        self.winner
    }
}
```

Reply:
I am also more inclined not to deal with mutating objects. In my exp, that often results in errors and more complex code.

Reply:
refactored according to example to remove code duplication

Reply:
I don't really see how turning `RawParticipant` into a `Participant` (i.e. adding the `winner` field) should not be considered mutating objects. Before Dusan introduced the accessor methods both solutions were equally prone to misuse because you could just change the public `winner` field of the `Participant` anyway.
Is the concept of a `RawParticipant` really so important that we should have a `domain` object for it?

I really want to avoid introducing more and more types which seem the same but are not quite the same. Otherwise you'll have the mental overhead of having to know which subtle detail is different in this particular version of the type.
WDYT about this to reduce the ways mutation can lead to issues (again I feel like what we did with the 2 types should also be considered mutating). That way you can only mutate the structs once in a predictable manner that makes sense in the domain:
```rust
#[derive(Clone)]
pub struct Participant {
    solution: Solution,
    driver: Arc<infra::Driver>,
    winner: bool,
}

impl Participant {
    pub fn driver(&self) -> &Arc<infra::Driver> {
        &self.inner.driver
    }

    pub fn solution(&self) -> &Solution {
        &self.inner.solution
    }

    pub fn is_winner(&self) -> bool {
        self.winner
    }
    
    pub fn mark_as_winner(&mut self) {
        self.is_winner = true;
    }
}
```

Reply:
If you are really adamant about type safety (i.e. having 2 types) I would at least make the names of the 2 structs more obvious: `UnrankedParticipant` and `(Ranked)Participant`.

Reply:
We could even play around with the type state pattern (although this is admittedly quite a lot of boiler plate code to achieve a simple thing):
```rust
struct Participant<State> {
    solution: Solution,
    driver: Arc<infra::Driver>,
    is_winner: bool,
    state: State,
}

struct Unranked;
struct Ranked;

impl <T> Participant<T> {
    pub fn solution(&self) -> &Solution {
        &self.solution
    }

    pub fn driver(&self) -> &Arc<infra::Driver> {
        &self.driver
    }
}

impl Participant<Unranked> {
    pub fn new(solution: Solution, driver: Arc<infra::Driver>) -> Self {
        Self {
            solution,
            driver,
            is_winner: false,
            state: Unranked
        }
    }

    pub fn rank(self, is_winner: bool) -> Participant<Ranked> {
        Participant::<Ranked> {
            is_winner,
            state: Ranked,
            solution: self.solution,
            driver: self.driver
        }
    }
}

impl Participant<Ranked> {
    pub fn is_winner(&self) -> bool {
        self.is_winner
    }
}
```

Reply:
The biggest concern would be if we use non-initialized `is_winner` while thinking it's initialized. I think I would be fine with it if the whole capturing of solve response and competition is done in a single step, for example, if we encapsulate everything in a `Competition::new()` constructor or something.

But I found you typestate pattern most adequate for this use case so I took it. 👍